### PR TITLE
Migrate to Gradle 7.0.2 and AGP 7.0.0

### LINKED ISSUE #87649: [flutter_tools] Support for Android Gradle Plugin 7.0.0
As Android Gradle Plugin 7.0.0 [released](https://developer.android.com/studio/releases/gradle-plugin#7-0-0)

### Steps to reproduce 

1.  Execute `flutter create myapp`
2.  Locate `myapp/android` and delete `gradle` distribution folder 
3. Update Android Gradle Plugin `myapp/android/build.gradle` to `7.0.0`
4.  Execute `flutter run -v` from the root directory of myapp.

### Actual Results:
Flutter doesn't support this version and throws the following exception: (using 7.0.2 default version) 
```console
[  +33 ms] Building APK
[  +40 ms] Running Gradle task 'assembleDebug'...
[  +15 ms] LocalFile: 'C:\Users\Taha\StudioProjects\triage-companion\android\build.gradle' provides AGP version: 7.0.0
[+9740 ms] Exception: Unsupported Android Plugin version: 7.0.0.
[   +2 ms] "flutter run" took 17,921ms.
```

The Source of this exception is [here](https://github.com/flutter/flutter/blob/c4a2a3e92d9b076cbcb7b76e7017410ddbc3bb5f/packages/flutter_tools/lib/src/android/gradle_utils.dart#L182-L186)

### Expected  Results:
The app runs with updated AGP and Gradle distribution is downloaded

### PR DESCRIPTION
**Pull request by TahaTesser**
- Migrated tools to Gradle 7.0.2 and AGP 7.0.0 (updated tests too).
- Migrated templates to Gradle 7.0.2 and AGP 7.0.0.
- Migrated flutter/dev samples to Gradle 7.0.2 and AGP 7.0.0.
- Migrated flutter/examples to Gradle 7.0.2 and AGP 7.0.0.
- Migrated packages/integration_test & packages/integration_test/example  to Gradle 7.0.2 and AGP 7.0.0.

Fixes https://github.com/flutter/flutter/issues/87649

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by TahaTesser** [comment]
@blasten 
Tests fails as Java 11 is  required for AGP 7.0.0+ 😅

**Comment by blasten** [comment]
@TahaTesser perfect timing. 

cc @CaseyHillers @godofredoc . Would it be possible to prioritize the work needed to get flutter_deps working on a PR like this?  e.g. parse .ci.yaml in LUCI and set the JAVA_HOME & PATH based on the content of this file.

**Comment by CaseyHillers** [comment]
> cc @CaseyHillers @godofredoc . Would it be possible to prioritize the work needed to get flutter_deps working on a PR like this? e.g. parse .ci.yaml in LUCI and set the JAVA_HOME & PATH based on the content of this file.

I could send a PR that changes it so presubmit uses the latest presubmit dependencies from ci.yaml. However, it would temporarily break tip of tree until the luci config change has propagated into prod. @blasten what is the current process for updating deps like this? Is it expected to be a hard break?

**Comment by godofredoc** [comment]
> t be possible to prio



> @TahaTesser perfect timing.
> 
> cc @CaseyHillers @godofredoc . Would it be possible to prioritize the work needed to get flutter_deps working on a PR like this? e.g. parse .ci.yaml in LUCI and set the JAVA_HOME & PATH based on the content of this file.

The only change required from the infra point of view is to update the version of the cipd dependency. @blasten are the new cipd packages ready?

**Comment by blasten** [comment]
Ideally, it's not a breaking change. This change primarily affects someone who creates a new app using `flutter create`. Ideally, we keep a test that uses the current version of Gradle and JAVA version.

**Comment by blasten** [comment]
> are the new cipd packages ready? 

Yes. However, @CaseyHillers mentioned that changes like this cannot be tested in the PR

**Comment by godofredoc** [comment]
> > t be possible to prio
> 
> > @TahaTesser perfect timing.
> > cc @CaseyHillers @godofredoc . Would it be possible to prioritize the work needed to get flutter_deps working on a PR like this? e.g. parse .ci.yaml in LUCI and set the JAVA_HOME & PATH based on the content of this file.
> 
> The only change required from the infra point of view is to update the version of the cipd dependency. @blasten are the new cipd packages ready?

Actually took a look and it seems like the new versions are there. The only change needed is to set "version": "version:11" in all the places using openjdk e.g. https://github.com/flutter/flutter/blob/master/.ci.yaml#L102 here is an example: https://github.com/flutter/engine/blob/master/.ci.yaml#L25

**Comment by blasten** [comment]
>  The only change needed is to set "version": "version:11" in all the places using openjdk 

@godofredoc I thought that changes aren't propagated automatically in the PR. Is that the case?   PRs like this that also have dependencies on the environment cannot be landed easily unless everything is pushed atomically.

**Comment by godofredoc** [comment]
> > The only change needed is to set "version": "version:11" in all the places using openjdk
> 
> @godofredoc I thought that changes aren't propagated automatically in the PR. Is that the case? PRs like this that also have dependencies on the environment cannot be landed easily unless everything is pushed atomically.

Post-submit like device lab won't be tested but as long as presubmit ones + manual executions of 1 linux, 1 mac and 1 windows device lab tests are successful we can have high confidence the new jdk will work properly.

**Comment by CaseyHillers** [comment]
I added some documentation in https://github.com/flutter/cocoon/pull/1321

From an infra perspective, it's easier to support bring up builders intended to validate new versions like this. There's a few race conditions that require more extensive work to iron out to support this functionality in presubmit.

**Comment by christopherfujino** [comment]
@CaseyHillers, @godofredoc, @blasten 
So what is the way forward to land this?

**Comment by blasten** [comment]
The current solution is described in https://github.com/CaseyHillers/cocoon/blob/1f1fa81e2588904908d613111a9d72e784cc6e24/CI_YAML.md#upgrading-dependencies

I wanted to try this by myself, and send a PR to see

**Comment by blasten** [comment]
@TahaTesser I migrated the complex_layout tests to Gradle/AGP 7.0. You can see the PR here: https://github.com/flutter/flutter/pull/88540

# Steps
1. Bump the AGP version in android/build.gradle
2. Bump the Gradle version in android/gradle/wrapper/gradle-wrapper.properties 
3. Cd into android and run ./gradlew generateLockfiles
4. Add the dependencies to each task in `.ci.yaml` See https://github.com/flutter/flutter/pull/88540 for reference.
5. Mark each test as flaky. (add bringup: true)
6. Get LGTM
7. Check the status of each task on https://flutter-dashboard.appspot.com/
8. If the tree got red because one of the task is not passing, then revert, fix, and reland.
9. If the tree is green, but one of the task flagged as flaky is not passing, then fix forward. (Send another PR with the fix)
10. Once all the tasks are passing, send another PR to remove bringup: true from the tasks in .ci.yaml.

**Comment by TahaTesser** [comment]
@blasten 
I have all updated samples, updated Gradle lock files, `ci.yaml` for all the ci tasks that I could find under `flutter/dev` and all the ci tasks that were failing here

Now only `ci_yaml flutter roller` fails for some reason.

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:1861]:
```suggestion
          {"dependency": "xcode"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:1913]:
```suggestion
          {"dependency": "goldctl"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:1938]:
```suggestion
          {"dependency": "goldctl"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:1963]:
```suggestion
          {"dependency": "goldctl"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:1988]:
```suggestion
          {"dependency": "goldctl"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:2119]:
```suggestion
          {"dependency": "gems"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:2144]:
```suggestion
          {"dependency": "gems"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:2169]:
```suggestion
          {"dependency": "gems"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:2194]:
```suggestion
          {"dependency": "gems"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:2220]:
```suggestion
          {"dependency": "gems"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:2246]:
```suggestion
          {"dependency": "gems"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:2297]:
```suggestion
          {"dependency": "gems"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:2348]:
```suggestion
          {"dependency": "gems"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:2405]:
```suggestion
          {"dependency": "goldctl"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:2434]:
```suggestion
          {"dependency": "goldctl"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:2463]:
```suggestion
          {"dependency": "goldctl"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:3663]:
```suggestion
          {"dependency": "vs_build"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:3687]:
```suggestion
          {"dependency": "vs_build"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:3711]:
```suggestion
          {"dependency": "vs_build"}
```

File: .ci.yaml
Comment (CaseyHillers) [inline .ci.yaml:4032]:
```suggestion
          {"dependency": "vs_build"}
```

**Comment by CaseyHillers** [comment]
Several of the blocks in ci.yaml have to be json parseable, so `Linux ci_yaml flutter roller` is ensuring all properties are json encoded. Usually this means a trailing comma. I reviewed the file and didn't find any other issues.

**Comment by TahaTesser** [comment]
@CaseyHillers 
Thanks for the review, I've removed such trailing commas, I searched for similar under caches and dependencies tags. Could not find any more of these but `ci_yaml flutter roller` still failing

**Comment by TahaTesser** [comment]
@CaseyHillers 
Found one more trailing comma. the test is in the queue, fingers crossed.

**Comment by TahaTesser** [comment]
@blasten 
All green, thanks to @CaseyHillers `ci_yaml flutter roller` is resolved

**Review by blasten — APPROVED** [review]
LGTM

**Comment by blasten** [comment]
@TahaTesser please fix the conflict, and I will merge once all checks are green.

**Comment by TahaTesser** [comment]
Closing this for a new PR, I will in a day or two

**Pull request closed without merging**
